src/platform/datapath_win.c

```diff
@@ -0,0 +1,615 @@
+    } else if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_XDP) {
+        RawSendDataFree(SendData);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+_Success_(return != NULL)
+QUIC_BUFFER*
+CxPlatSendDataAllocBuffer(
+    _In_ CXPLAT_SEND_DATA* SendData,
+    _In_ uint16_t MaxBufferLength
+    )
+{
+    if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_USER) {
+        return SendDataAllocBuffer(SendData, MaxBufferLength);
+    } else if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_XDP) {
+        return RawSendDataAllocBuffer(SendData, MaxBufferLength);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+    return NULL;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatSendDataFreeBuffer(
+    _In_ CXPLAT_SEND_DATA* SendData,
+    _In_ QUIC_BUFFER* Buffer
+    )
+{
+    if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_USER) {
+        SendDataFreeBuffer(SendData, Buffer);
+    } else if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_XDP) {
+        RawSendDataFreeBuffer(SendData, Buffer);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+BOOLEAN
+CxPlatSendDataIsFull(
+    _In_ CXPLAT_SEND_DATA* SendData
+    )
+{
+    if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_USER) {
+        return SendDataIsFull(SendData);
+    } else if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_XDP) {
+        return RawSendDataIsFull(SendData);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+    return FALSE;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+QUIC_STATUS
+CxPlatSocketSend(
+    _In_ CXPLAT_SOCKET* Socket,
+    _In_ const CXPLAT_ROUTE* Route,
+    _In_ CXPLAT_SEND_DATA* SendData
+    )
+{
+    if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_USER) {
+        return SocketSend(Socket, Route, SendData);
+    } else if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_XDP) {
+        return RawSocketSend(CxPlatSocketToRaw(Socket), Route, SendData);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+    return QUIC_STATUS_NOT_SUPPORTED;
+}
+
+void
+CxPlatDataPathProcessCqe(
+    _In_ CXPLAT_CQE* Cqe
+    )
+{
+    // TODO: refactoring
+    switch (CxPlatCqeType(Cqe)) {
+    case CXPLAT_CQE_TYPE_SOCKET_IO: {
+        DATAPATH_IO_SQE* Sqe =
+            CONTAINING_RECORD(CxPlatCqeUserData(Cqe), DATAPATH_IO_SQE, DatapathSqe);
+        if (Sqe->IoType == DATAPATH_XDP_IO_RECV || Sqe->IoType == DATAPATH_XDP_IO_SEND) {
+            RawDataPathProcessCqe(Cqe);
+        } else {
+            DataPathProcessCqe(Cqe);
+        }
+        break;
+    }
+    case CXPLAT_CQE_TYPE_SOCKET_SHUTDOWN: {
+        RawDataPathProcessCqe(Cqe);
+        break;
+    }
+    default: CXPLAT_DBG_ASSERT(FALSE); break;
+    }
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicCopyRouteInfo(
+    _Inout_ CXPLAT_ROUTE* DstRoute,
+    _In_ CXPLAT_ROUTE* SrcRoute
+    )
+{
+    if (!IS_LOOPBACK(SrcRoute->RemoteAddress)) {
```

Comment:
>!IS_LOOPBACK(SrcRoute->RemoteAddress) [](http://example.com/codeflow?start=8&length=37)

How is the route info copy logic coupled to loopback?

Reply:
loopback is not enough to determine the logic. adding one more parameter